File: 김수미/boj/G5_9205.py
''' 
* Author    : Kim Su Mi
 * Date      : 2024.09.26(Tue)
 * Runtime   : 52 ms
 * Memory    : 34088 KB
 * Algorithm : BFS
'''

import sys
from collections import deque 

def bfs():
    queue = deque()
    queue.append((home_x, home_y))

    while queue:
        # 편의점을 들리지않고, 바로 페스티벌로 갈 수 있다면
        x, y = queue.popleft()
        if abs((festival_x - x)) + abs((festival_y - y)) <= 1000:
            print("happy")
            return

        # 편의점을 들러야한다면 모든 편의점 방문한 후
        for i in range(n):
            # 그 편의점을 방문하지않았다면
            if not visited[i]:
                new_x, new_y = convenience_stores[i]
                if abs((new_x - x)) + abs((new_y - y)) <= 1000:
                    visited[i] = 1
                    queue.append((new_x, new_y))

    print("sad")
    return

t = int(sys.stdin.readline())

for _ in range(t):
    n = int(sys.stdin.readline())
    convenience_stores = []

    # 시작 지점은 정해져있음
    home_x, home_y = map(int, sys.stdin.readline().split())
    for _ in range(n):
        convenience_stores.append(tuple(map(int, sys.stdin.readline().split())))

    festival_x, festival_y = map(int, sys.stdin.readline().split())

    # 방문한 편의점 확인
    visited = [0 for _ in range(n+1)]
    bfs()

File: 김수미/boj/G5_9205.java
/**
 * Author    : Kim Su Mi
 * Date      : 2024.09.26(Tue)
 * Runtime   : 128 ms
 * Memory    : 14928 KB
 * Algorithm : BFS
 */

package 김수미.boj;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;
import java.util.StringTokenizer;

public class G5_9205 {

	static int t, n, homeX, homeY, festivalX, festivalY;
	static List<int[]> convenienceStores;
	static int[] visited;

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
		StringTokenizer st = null;
		t = Integer.parseInt(br.readLine());

		for (int i = 1; i <= t; i++) {
			n = Integer.parseInt(br.readLine());
			convenienceStores = new ArrayList<>();

			st = new StringTokenizer(br.readLine());
			homeX = Integer.parseInt(st.nextToken());
			homeY = Integer.parseInt(st.nextToken());

			for (int j = 0; j < n; j++) {
				st = new StringTokenizer(br.readLine());
				convenienceStores.add(new int[]{Integer.parseInt(st.nextToken()), Integer.parseInt(st.nextToken())});
			}

			st = new StringTokenizer(br.readLine());
			festivalX = Integer.parseInt(st.nextToken());
			festivalY = Integer.parseInt(st.nextToken());

			visited = new int[n];
			Arrays.fill(visited, 0);

			bw.write(bfs() ? "happy\n" : "sad\n");
		}

		// bw flush, close 해줘야함.
		bw.flush();
		bw.close();
	}

	static boolean bfs() {
		Queue<int[]> q = new LinkedList<>();
		q.add(new int[]{homeX, homeY});

		while (!q.isEmpty()) {
			int[] cur = q.poll();
			int x = cur[0], y = cur[1];
			if (Math.abs(festivalX - x) + Math.abs(festivalY - y) <= 1000) {
				return true;
			}

			for (int i=0; i<n ; i++){
				if (visited[i] == 0) {
					int newX = convenienceStores.get(i)[0], newY = convenienceStores.get(i)[1];
					if (Math.abs(newX - x) + Math.abs(newY - y) <= 1000) {
						visited[i] = 1;
						q.add(new int[]{newX, newY});
					}
				}
			}
		}
		return false;
	}

}
